r/R/trackplots.R

```diff
@@ -450,6 +478,126 @@ trackplot_gene <- function(transcripts, region, exon_size = 2.5, gene_size = 0.5
+  
+  # if data is empty, return data or empty plot
+  if (nrow(loci) == 0) {
+    if (return_data) {
+      return(loci)
+    } else {
+      return(trackplot_empty(region, track_label))
+    }
+  }
+  data <- trackplot_calculate_segment_height(data)
+  # Adjust the endpoints of any partially overlapping elements to fit within
+  # the plot boundaries
+  data <- dplyr::mutate(
+    data,
+    start = pmax(region$start, pmin(region$end, start)),
+    end = pmax(region$start, pmin(region$end, end)),
+    facet_label = track_label
+  )
+
+  plot <- ggplot2::ggplot(data, ggplot2::aes(x = start, xend = end, y =y, yend = y)) + 
+    ggplot2::geom_segment(linewidth=annotation_size)  + 
+    ggplot2::scale_x_continuous(limits = c(region$start, region$end), expand = c(0, 0), labels=scales::label_number()) +
+    ggplot2::scale_y_discrete(labels = NULL, breaks = NULL) +
+    ggplot2::labs(x = "Genomic Position (bp)", y = NULL) + 
+    ggplot2::guides(size="none", linewidth="none") +
+    ggplot2::facet_wrap("facet_label", strip.position="left") +
+    trackplot_theme()
+  # add labels if given
+  if (!is.null(label_by)) {
+    plot <- plot +
+      ggrepel::geom_text_repel(
+        data = data,
+        ggplot2::aes(label = !!as.symbol(label_by)),
+        size = label_size,
+        position = ggrepel::position_nudge_repel(y = 0.25)
+      )
+  }
+  # add in arrows
+  if (!is.null(direction_by)) {
+    arrows <- trackplot_create_arrow_segs(data, region, 150, c(direction_by, "y"))
+    plot <- plot + 
+      ggplot2::geom_segment(arrows, mapping=aes(
+        x = ifelse(!!as.symbol(direction_by) == TRUE, start, end),
+        xend = ifelse(!!as.symbol(direction_by) == TRUE, end, start),
+        y = y, yend = y,
+      ),
+      arrow=grid::arrow(length=grid::unit(annotation_size/sin(arrow_angle*pi/180)*.3, "mm"), angle = arrow_angle),
+      colour = "black",)
+  }
+  # Route to different colouring aesthetic paths depending on color_by and color type/presence
+  if (!is.null(color_by)) {
+    plot <- plot + ggplot2::aes(color = .data[[color_by]])
+    if (is.character(data[[color_by]]) || is.factor(data[[color_by]])) {
+      if (!is.null(colors)) {
+        assert_true(length(colors) >= length(unique(data[[color_by]])))
+        plot <- plot + ggplot2::scale_color_manual(values = colors)
+      } else {
+        plot <- plot + ggplot2::scale_color_manual(values = discrete_palette("tableau", length(unique(data[[color_by]]))))
+      }
+    } else if (is.numeric(data[[color_by]])) {
+      if (!is.null(colors)) {
+        assert_true(length(colors) >= 2,
+                  "Length of colors must be greater than or equal to 2")
+        plot <- plot + ggplot2::scale_color_gradientn(colors=colors)
+      } else {
+        plot <- plot + ggplot2::scale_color_gradient(
+          high = munsell::mnsl("5P 3/12"), 
+          low = munsell::mnsl("5P 7/12"),
+        )
+      }
+    }
+  }
```

Comment:
A few suggestions:
1. Replace the `munsell::mnsl` colors with their hex codes -- don't want to take a new library dependency. (Though fine to put the `mnsl` call as a comment to explain how you got the colors)
2. If you need some `assert_true` calls, do them at the top of the function before work is performed -- just general BPCells policy to avoid users waiting through a potentially slow function call before getting an error.
3. For discrete colors, you can just run `as.factor` when constructing the `data` tibble so later on in the function you don't need to worry about character values

Last suggestion is more minor code style, but I think you can slightly simplify if you use `scale_color_gradientn` in the default color scale as well.
```r
if (!is.null(color_by)) {
    plot <- plot + ggplot2::aes(color = color)
    if (is.factor(data$color)) {
      if (is.null(colors)) colors <- discrete_palette("tableau", length(levels(data$color)))
      plot <- plot + ggplot2::scale_color_manual(values = colors)
    } else {
      # default continuous palette `c(munsell::mnsl("5P 3/12"), munsell::mnsl("5P 7/12"))`
      if (!is.null(colors)) colors <- c("#D099F2", "#6D2884")
      plot <- plot + ggplot2::scale_color_gradientn(colors=colors)
    }
  }
```